scenario_player/runner.py

```diff
@@ -153,289 +335,326 @@ def local_seed(self) -> str:
             seed = seed_file.read_text().strip()
         return seed
 
-    def select_chain(self, chain_urls: Dict[str, List[str]]) -> Tuple[str, List[str]]:
-        """Select a chain and return its name and RPC URL.
+    def ensure_token_network_discovery(
+        self, token: CustomToken, token_network_addresses: TokenNetworkAddress
+    ) -> None:
+        """Ensure that all our nodes have discovered the same token network."""
+        for node in self.node_controller:  # type: ignore
+            node_endpoint = API_URL_TOKEN_NETWORK_ADDRESS.format(
+                protocol=self.protocol,
+                target_host=node.base_url,
+                token_address=to_checksum_address(token.address),
+            )
+            address = wait_for_token_network_discovery(
+                node_endpoint, self.definition.settings, self.session
+            )
+            if to_canonical_address(address) != Address(token_network_addresses):
+                raise RuntimeError(
+                    f"Nodes diverged on the token network address, there should be "
+                    f"exactly one token network available for all nodes. Current "
+                    f"values : {to_hex(token_network_addresses)}"
+                )
+
+    def run_scenario(self) -> None:
+        with Janitor() as nursery:
+            self.node_controller = NodeController(self, self.definition.nodes, nursery)
```

Comment:
A bit nitpicky: I'm not a fan of creating instance variables outside of the call scope of `__init__`. That's just asking for `AttributeError`s later on.

Reply:
Agreed, I don't like partially initialized objects. I did not try to fix that though, but IMO this should go in a follow up.